src/Models.py

```diff
@@ -0,0 +1,155 @@
+# File for Layer and Model implementations
+import torch
+
+from torch.nn import Linear, ReLU, BatchNorm1d, Module, Sequential
+from torch_geometric.nn import MessagePassing, global_mean_pool
+from torch_scatter import scatter
+
+class EGNNMPNNLayer(MessagePassing):
+    def __init__(self, emb_dim=64, edge_dim=32, aggr='add'):
+        """EGNN Message Passing Neural Network Layer
+
+        This layer is equivariant in 3D rotations and translations and permutations with respect to the input. 
+
+        Args:
+            emb_dim: (int) - hidden dimension `d`
+            edge_dim: (int) - edge feature dimension `d_e`
+            aggr: (str) - aggregation function `\oplus` (sum/mean/max)
+        """
+        super().__init__(aggr=aggr)
+
+        self.emb_dim = emb_dim
+        self.edge_dim = edge_dim
+
+        self.mlp_msg = self.mlp_msg = Sequential(
+            Linear(2*emb_dim + edge_dim + 1, emb_dim), BatchNorm1d(emb_dim), ReLU(),
+            Linear(emb_dim, emb_dim), BatchNorm1d(emb_dim), ReLU()
+          )  
+        
+        self.mlp_upd_features = Sequential(
+            Linear(2*emb_dim, emb_dim), BatchNorm1d(emb_dim), ReLU(), 
+            Linear(emb_dim, emb_dim), BatchNorm1d(emb_dim), ReLU()
+          )
+        self.mlp_upd_coord = Sequential(
+            Linear(2*emb_dim + edge_dim + 1, emb_dim), BatchNorm1d(emb_dim), ReLU(), 
+            Linear(emb_dim, 1), BatchNorm1d(1), ReLU()
+          )
+
+    def forward(self, h, pos, edge_index, edge_s):
+        """
+        The forward pass updates node features `h` via one round of message passing.
+
+        Args:
+            h: (n, d) - initial node features
+            pos: (n, 3) - initial node coordinates
+            edge_index: (e, 2) - pairs of edges (i, j)
+            edge_s: (e, d_e) - edge features
+
+        Returns:
+            out: (n, d) - updated node features
+        """
+        out = self.propagate(edge_index, h=h, edge_s=edge_s, pos=pos)
+        return out
+     
+    def message(self, h_i, h_j, pos_i, pos_j, edge_s):
+      """The `message()` function constructs messages from source nodes j 
+       to destination nodes i for each edge (i, j) in `edge_index`.
+    
+       Args:
+           h_i: (e, d) - destination node features
+           h_j: (e, d) - source node features
+           pos_i: (e, 3) - destination node coordinates
+           pos_j: (e, 3) - source node coordinates
+           edge_s: (e, d_e) - edge features
+        
+       Returns:
+           msg: (e, d) - messages `m_ij` passed through MLP `\psi`
+      """ 
+      msg = torch.concat([h_i, h_j, (pos_i-pos_j).norm(dim=1, p=2).reshape(-1,1), edge_s], dim=-1)
+      return (self.mlp_msg(msg), self.mlp_upd_coord(msg)*(pos_i-pos_j))
```

Comment:
Again, for readability, make sure to not pack too much complexity into a single line - this is just about the maximum of complexity I would put in a single line

Also, probably it would make sense to rename `mlp_upd_coord` to `mlp_coord_msg` and `mlp_msg` to `mlp_feat_msg` or so?